### DISCUSSION TITLE
Authorization and registration via api

### DISCUSSION THREAD

**Discussion by mr-scrpt** [nextauthjs/next-auth#10192] [Help]
Hey, everybody. There is a site on nextjs 14, it is a training project on which I am practicing - it is an admin panel for online store, in it you can create categories, products, etc - in general it is not very important. But also in this admin panel there is registration via next-auth, now two providers are connected - email and github - everything works as it should. But this admin panel by idea gives outward api to which the front-end client of the online store will address, for which the admin panel is made, that is two different services - admin panel with api and client which will use the provided api. Here I have a question - I need to register users from the client via api, that is, in the online store will be a form in which you can enter a login and password and send these data via api, I can do it with the next auth? Perhaps someone can tell me how to implement this correctly?

**Comment by ananuness** [comment]
Yes! You can send the email and password for example to your Next API and then the Next auth will do the job, here is a example:

First, the versions I used:

```json
{
  "@auth/prisma-adapter": "^1.0.10",
  "next": "13.4.2", // you can use 14
  "next-auth": "4.22.1",
}
```


### Authorization

In your `authOptions`, it's where you will configure everything!

```ts
import { compare } from "bcryptjs";
import { NextAuthOptions } from "next-auth"; // this one is just if you're using typescript
import { PrismaAdapter } from "@auth/prisma-adapter"; // you will need a corresponding adapter for your DB implementation
import CredentialsProvider from "next-auth/providers/credentials"; // as I'm using email/password login, I need to import this provider
import prisma from "./prisma";

export const authOptions: NextAuthOptions = {
  /**
   * @see https://github.com/prisma/prisma/issues/16117
   * @ts-ignore
   */
  adapter: PrismaAdapter(prisma as any),
 // in this array you inform which providers you will use, you can inform a Google, Github, Twitter and many other ways to login your user
  providers: [
    CredentialsProvider({
      name: "Credentials",
      credentials: { // this one you use if you don't have any login page implemented
        email: {
          label: "E-mail",
          type: "text",
          placeholder: "Your e-mail",
        },
        password: {
          label: "Password",
          type: "password",
          placeholder: "Your password",
        },
      },
      async authorize(credentials): Promise<any> { // here is were you will define YOUR logic when the user tries to log in
        const user = await prisma.user.findUnique({
          where: {
            email: credentials?.email,
          },
        });

        if (!user || !credentials?.password) {
          throw new Error("E-mail e/ou senha invalidos");
        }

        const matchPassword = await compare(
          credentials.password,
          user.hashedPassword,
        );

        if (!matchPassword) throw new Error("Invalid e-mail or password");

        return {
          ...user,
          name: `${user.name} ${user.lastName}`, // here is just my logic, you don't need to do the same thing
        };
      },
    }),
  ],
  session: {
    strategy: "jwt", // you NEED to define this strategy if you're using JWT token, because the next-auth use database by default
  },
  pages: {
    signIn: "/login", // as I have a login page implemented, I just passed the route here to replace the default one!
  },
  callbacks: {
    // here I defined this callback because I needed the id in the session to my app, you can implement and store things here!
    async session({ session, token }) {
      session.user.id = token.sub;

      return session;
    },
  },
};
```

Next, in the `/api/auth/[...nextauth]`, you will just expose your handler (your logic) for the Next backend:

```ts
import NextAuth from "next-auth";
import { authOptions } from "@/config/authOptions";

const handler = NextAuth(authOptions);

export { handler as GET, handler as POST };
```

Also, as I was using typescript in the `authOptions` I needed to declare the next-auth module to add the ID I wanted to the session storage:

```ts
// next-auth.d.ts
import NextAuth, { DefaultSession } from "next-auth";

declare module "next-auth" {
  interface Session extends DefaultSession {
    user: {
      id?: string;
    } & DefaultSession["user"];
  }
}
```
This one can be really useful if you need more than the default session which is `name, email and image`!

And finally, here is a example where I call it in my client:

```ts
// I was using the react-hook-form, but you can call it normally
const handleSubmit = hookFormSubmit(async ({ email, password }) => {
     // and just like this, you can make a request to your logic defined in the authorize method in authOptions
    const response = await signIn("credentials", {
      email,
      password,
      redirect: false, // this one was defined just because I wanted to do more before redirect!
      callbackUrl: "/welcome", // and here is were I wanted to go after everything
    });

    if (response?.error) {
      setLoginApiError(response.error);
    } else {
      setLoginApiError(null);
    }
  });
```

And remember to wrap your whole app (in your RootLayout) with the `SessionProvider`! There's all those details, but it's kinda easy to set it up!

### Registration

For the registration I implemented it using the next backend only:

```ts
// /api/register
import { hash } from "bcryptjs";
import { NextRequest } from "next/server";
import { registerSchema, RegisterData } from "@/register/registerSchema";
import { sendResponse } from "@/helpers/api-response"; // this is just a helper to the default NextResponse.json()
import prisma from "@/config/prisma";

export async function POST(request: NextRequest) {
  try {
    const body = (await request.json()) as RegisterData;
    const { name, lastName, email, password } = registerSchema.parse(body);

    const hashedPassword = await hash(password, 12);

    const response = await prisma.user.create({
      data: {
        name,
        lastName,
        email,
        hashedPassword,
        image: null,
      },
    });

    return sendResponse(201, response);
  } catch (error: any) {
    if (error.code === "P2002") {
      return sendResponse(409, "E-mail already in use");
    }

    return sendResponse(
      500,
      "Ops, something happened! Try again later",
    );
  }
}
```

I will leave some links where you can find more about it, and you can always find a good tutorial on youtube! I hope it helps and don't hesitate to ask!

### Links
- [Next auth client API](https://next-auth.js.org/getting-started/client): here you will see everything they offer to you access sessions, providers, signin and signout methods in the client
- [Next auth REST API](https://next-auth.js.org/getting-started/rest-api): here you will find all the routes they provide to you!
- [Next auth with typescript](https://next-auth.js.org/getting-started/typescript): this one will help you with the types and interfaces
- [Next auth adapters](https://next-auth.js.org/v3/adapters/overview): an Adapter in NextAuth.js connects your application to whatever database or backend system you want to use
- [My demo using Next auth with credentials and Google/Github providers](https://github.com/ananuness/next-auth-demo/): it's a demo to set it up and showing the flow, but the readme is in portuguese, I will translate it soon, but you can use a translator while I add it.
- [A good tutorial for Next auth](https://www.youtube.com/watch?v=2kgqPvs0j_I&pp=ugMICgJwdBABGAHKBQluZXh0IGF1dGg%3D): you can search about it using a adapter of your choice too! This one is with prisma
